kafka-handler/src/java/org/apache/hadoop/hive/kafka/KafkaSerDe.java

```diff
@@ -369,6 +379,20 @@ private SubStructObjectInspector(StructObjectInspector baseOI, int toIndex) {
     }
   }
 
+  static class ConfluentAvroBytesConverter extends AvroBytesConverter {
+    ConfluentAvroBytesConverter(Schema schema) {
+      super(schema);
+    }
+
+    @Override
+    Decoder getDecoder(byte[] value) {
+      /**
+       * Confluent 4 magic bytes that represents Schema ID as Integer. These bits are added before value bytes.
+       */
+      return DecoderFactory.get().binaryDecoder(value, 5, value.length - 5, null);
```

Comment:
I'm not immidatiately aware of the parameters here, but assuming this operates like a byte-slice with start and end indicies, why `value.length - 5`?

+ Why not include the `CachedSchemaRegistryClient` class and actually **get** the schema using the ID?

Reply:
> Why not include the `CachedSchemaRegistryClient` class and actually **get** the schema using the ID?

Great Idea, would love to see this coming soon where we can use `SchemaRegistryClient` to import all the existing schemas use it as part of the query compile time and ship it to tasks via the serialized plan.

Reply:
One potential problem with getting the ID is that it could potentially change for each and every message in the topic. And the schema registry isn't guaranteed to preserve backward compatibility-ness against all topic data. I guess that's an implementation detail, but I wonder how Hive would handle that when building the ResultSet for altering schemas

Reply:
The problem is you have to provide schema during `initialize` and there is no record to read schema ID from at that time. I was thinking about switch `bytesConverter` to a function/supplier or factory and give it to the KafkaWritable and resolve byteconverter lazy way. However I'm quite new to Hive and need to deep more to mechanics how the low level api works. I'd like to see hive download schemas from schema registry, but I need to think about concept first.

About the compatibility, there are several compatibility modes in schema registry. It can be set for each topic and you can set BACKWARD compat for example.

And yea you can have objects with literally different schemas not just different versions in one topic and I think this problem Hive cannot address.

Reply:
That `value.length - 5` is cuz u provide `start` and `length` not `start` and `end` offsets.

Reply:
> About the compatibility, there are several compatibility modes in schema registry. It can be set for each topic and you can set BACKWARD compat

Right, but Hive will likely only work properly if the config really is set to backwards, or at the least, the schema provided can read all the data for the offsets that are provided. 

> with literally different schemas

Maybe if more schema data was exposed, this would be easier to handle? For example, the Avro namespace + top-level record name?

Reply:
> I'd like to see hive download schemas from schema registry

Linked to earlier, the `/schema` endpoint should be able to get the schema text, however I suspect that it'll cause an http request per Hive "stage"? For example, N requests if reading from N Kafka partitions? And more requests if trying to do fancy selects, unions, and joins?

Reply:
Well, I think schema registry is out of scope. It should be address in another PR.